File: backend/app/routes/notification.py
from flask import request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from . import api_bp
from .. import db
from ..models import Notification, User


@api_bp.route('/notifications', methods=['GET'])
@jwt_required()
def get_notifications():
    user_id = int(get_jwt_identity())
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 20, type=int)
    unread_only = request.args.get('unread_only', 'false').lower() == 'true'

    query = Notification.query.filter_by(user_id=user_id)
    
    if unread_only:
        query = query.filter_by(is_read=False)
    
    pagination = query.order_by(Notification.created_at.desc()).paginate(
        page=page, per_page=per_page, error_out=False
    )

    return jsonify({
        'notifications': [n.to_dict() for n in pagination.items],
        'total': pagination.total,
        'unread_count': Notification.query.filter_by(user_id=user_id, is_read=False).count(),
        'pages': pagination.pages,
        'current_page': page
    })


@api_bp.route('/notifications/unread-count', methods=['GET'])
@jwt_required()
def get_unread_count():
    user_id = int(get_jwt_identity())
    
    count = Notification.query.filter_by(user_id=user_id, is_read=False).count()
    
    return jsonify({
        'unread_count': count
    })


@api_bp.route('/notifications/<int:notification_id>/read', methods=['PUT'])
@jwt_required()
def mark_notification_read(notification_id):
    user_id = int(get_jwt_identity())
    
    notification = Notification.query.filter_by(id=notification_id, user_id=user_id).first()
    
    if not notification:
        return jsonify({'error': 'Notification not found'}), 404
    
    notification.is_read = True
    db.session.commit()
    
    return jsonify({
        'message': 'Notification marked as read',
        'notification': notification.to_dict()
    })


@api_bp.route('/notifications/read-all', methods=['PUT'])
@jwt_required()
def mark_all_notifications_read():
    user_id = int(get_jwt_identity())
    
    Notification.query.filter_by(user_id=user_id, is_read=False).update({'is_read': True})
    db.session.commit()
    
    return jsonify({
        'message': 'All notifications marked as read'
    })


@api_bp.route('/notifications/<int:notification_id>', methods=['DELETE'])
@jwt_required()
def delete_notification(notification_id):
    user_id = int(get_jwt_identity())
    
    notification = Notification.query.filter_by(id=notification_id, user_id=user_id).first()
    
    if not notification:
        return jsonify({'error': 'Notification not found'}), 404
    
    db.session.delete(notification)
    db.session.commit()
    
    return jsonify({
        'message': 'Notification deleted'
    })


def create_notification(user_id, type, title, content=None, related_id=None, related_type=None):
    notification = Notification(
        user_id=user_id,
        type=type,
        title=title,
        content=content,
        related_id=related_id,
        related_type=related_type
    )
    db.session.add(notification)
    db.session.commit()
    return notification

File: pbl6/src/stores/notification.js
import { defineStore } from 'pinia'
import request from '@/utils/request'

export const useNotificationStore = defineStore('notification', {
  state: () => ({
    notifications: [],
    unreadCount: 0,
    total: 0,
    pages: 0,
    currentPage: 1
  }),
  
  actions: {
    
    async fetchNotifications(page = 1, perPage = 20, unreadOnly = false) {
      try {
        const params = { page, per_page: perPage }
        if (unreadOnly) params.unread_only = true
        
        const data = await request.get('/notifications', { params })
        
        this.notifications = data.notifications
        this.total = data.total
        this.unreadCount = data.unread_count
        this.pages = data.pages
        this.currentPage = data.current_page
        
        return data
      } catch (error) {
        throw error
      }
    },
    
    async fetchUnreadCount() {
      try {
        const data = await request.get('/notifications/unread-count')
        
        this.unreadCount = data.unread_count
        return data.unread_count
      } catch (error) {
        throw error
      }
    },
    
    async markAsRead(notificationId) {
      try {
        const data = await request.put(`/notifications/${notificationId}/read`)
        
        const index = this.notifications.findIndex(n => n.id === notificationId)
        if (index !== -1) {
          this.notifications[index].is_read = true
        }
        this.unreadCount = Math.max(0, this.unreadCount - 1)
        
        return data
      } catch (error) {
        throw error
      }
    },
    
    async markAllAsRead() {
      try {
        const data = await request.put('/notifications/read-all')
        
        this.notifications.forEach(n => n.is_read = true)
        this.unreadCount = 0
        
        return data
      } catch (error) {
        throw error
      }
    },
    
    async deleteNotification(notificationId) {
      try {
        const data = await request.delete(`/notifications/${notificationId}`)
        
        this.notifications = this.notifications.filter(n => n.id !== notificationId)
        
        return data
      } catch (error) {
        throw error
      }
    }
  }
})